### PR TITLE
Add Sourcegraph links for metrics not indexed by Searchfox

### LINKED ISSUE #404: Use sourcegraph to search for code in repos not indexed by searchfox
In #339 @Iinh added support for searching for glean metric definitions using searchfox. However, not everything we care about is currently indexed by searchfox.

Seperately, @Iinh found a great service called sourcegraph which indexes at least some of these repositories. For example, mozregression:

https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/mozilla/mozregression%24+&patternType=literal

For these cases, we should provide links to sourcegraph. 

This is on the complex end of things for a good first bug, but should be doable. Fixing this issue would involve:

* Looking at how @Iinh implemented things in #339
* Implementing similar functionality for getting links to sourcegraph (note we would *only* return links to sourcegraph where we do not have a searchfox link, since searchfox is more familiar and probably better for those cases) and putting them in the metadata table.

### PR DESCRIPTION
**Pull request by jcads**
Fixes #404 

### Pull Request checklist

- [x] The pull request has a descriptive title (and a reference to an issue it
      fixes, if applicable)
- [x] All tests and linter checks are passing
- [x] The pull request is free of merge conflicts

### CODE REVIEW

File: src/formatters/searchfox.js [hunk 1]
```diff
@@ -30,7 +30,21 @@ const searchfoxMap = {
   reference_browser: "reference-browser",
 };
 
-export const getSearchfoxLink = (app, metric) => {
+const sourcegraphMap = {
```
Comment (wlach) [inline src/formatters/searchfox.js:33]: Could you rename this file to something like codesearch.js?

File: src/formatters/searchfox.js [hunk 2]
```diff
@@ -30,7 +30,21 @@ const searchfoxMap = {
...
+
+  mozphab: "mozphab",
+
+  mozregression: "mozregression",
+};
+
+export const getMetricLink = (app, metric) => {
   const [category, name] = metric.split(/\.(?=[^.]+$)/);
 
   // Generate all possible casings for a Searchfox query:
```
Comment (wlach) [inline src/formatters/searchfox.js:50]:
```suggestion
  // Generate all possible casings for a search query:
```

File: src/formatters/searchfox.js [hunk 3]
```diff
@@ -63,5 +77,9 @@ export const getSearchfoxLink = (app, metric) => {
       ? `https://searchfox.org/mozilla-central/search?q=${allLanguagePatterns}&regexp=true`
       : `https://searchfox.org/mozilla-mobile/search?q=${allLanguagePatterns}&path=${searchfoxMap[app]}&regexp=true`;
   }
-  return undefined;
+
+  return (
```
Comment (wlach) [inline src/formatters/searchfox.js:81]: Let's check to make sure there is an entry for this app in sourcegraph, just as we do above for searchfox.

There may be some cases where we don't have a code search service for a repository.

File: src/formatters/searchfox.js [hunk 4]
```diff
@@ -30,7 +30,21 @@ const searchfoxMap = {
...
+  lockwise_ios: "lockwise-ios",
+
+  mach: "mach",
+
+  mozphab: "mozphab",
+
+  mozregression: "mozregression",
+};
+
+export const getMetricLink = (app, metric) => {
```
Comment (wlach) [inline src/formatters/searchfox.js:47]:
```suggestion
export const getCodeSearchLink = (app, metric) => {
```

File: src/data/schemas.js [hunk 5]
```diff
@@ -64,7 +64,7 @@ export const METRIC_DEFINITION_SCHEMA = [
     helpText:
       "Finds uses of this metric using simple matching. Metrics used dynamically may not appear in the results, only metrics defined in the application will be found.",
     linkFormatter: (metricId, appName) => {
-      return getSearchfoxLink(appName, metricId);
+      return getMetricLink(appName, metricId);
```
Comment (wlach) [inline src/data/schemas.js:67]: Could you rename this entry to `Code Search`? Seems more intuitive to me.

**Review by wlach — CHANGES_REQUESTED** [review]
Thanks @jcads! Just have a few requests for changes before this lands.

**Review by wlach — APPROVED** [review]
Thanks for this!

The only bug I can see is that it seems like we're still showing the "code search" option (with no link) even when it's unavailable. e.g. https://deploy-preview-468--glean-dictionary-dev.netlify.app/apps/bergamot/metrics/errors_marian

image

However that seems like an old issue, so I'm going to merge this as-is. 

@jcads would you be interested in investigating that as a followup?

**Comment by jcads** [comment]
You're welcome, @wlach! :tada: 

As for the bug, I'll gladly look into that.

**Comment by jcads** [comment]
I did some digging and I think it's because `bergamot` has just recently been added to https://probeinfo.telemetry.mozilla.org/v2/glean/app-listings, where the glean-dictionary data is pulled from. Therefore, it wasn't added to `codesearch.js` to be indexed.

And just a sidenote: bergamot can't be currently indexed on both searchfox and sourcegraph, so adding it in `codesearch.js` will not fix the bug.

Should the "code search" option be removed when the link is unavailable?

**Comment by wlach** [comment]
> And just a sidenote: bergamot can't be currently indexed on both searchfox and sourcegraph, so adding it in `codesearch.js` will not fix the bug.

Right :) I asked about this on the #github-admin channel on Matrix today. I think searchgraph is not yet indexing the `mozilla-extensions` organization. I can ping them later about this.
 
> Should the "code search" option be removed when the link is unavailable?

I think it would be better to just say something like `Unavailable: no search index specified for this application.`. That way people know the reason why it's missing.

**Comment by jcads** [comment]
> I think it would be better to just say something like Unavailable: no search index specified for this application.. That way people know the reason why it's missing.

Got it! Opening a new issue.

**Pull request merged**